queries.js

```diff
@@ -145,291 +145,250 @@ function parseShesmuType(type) {
     } else if (permitted_parameters[name].required) {
-      res
-        .status(400)
-        .send('Report requires ' + name + ', but paramter not supplied.');
-      return 0;
+      throw new ValidationError(
+        'Report requires ' + name + ', but it was not supplied.'
+      );
     }
   }
-  var providedNames = Object.keys(req.body.parameters);
-  for (nameIndex = 0; nameIndex < providedNames.length; nameIndex++) {
-    name = providedNames[nameIndex];
+  const providedNames = Object.keys(req.body.parameters);
+  for (let nameIndex = 0; nameIndex < providedNames.length; nameIndex++) {
+    const name = providedNames[nameIndex];
     if (!permitted_parameters.hasOwnProperty(name)) {
-      res.status(400).send('Invalid parameter ' + name + ' provided.');
-      return 0;
+      throw new ValidationError('Invalid parameter ' + name + ' provided.');
     }
   }
   return 1;
 }
 
 // add query functions
 
-function getAllReports(req, res, next) {
-  db
-    .any('select * from report')
-    .then(function(data) {
-      res.status(200).json(data);
-    })
-    .catch(function(err) {
-      return next(err);
-    });
+async function getAllReports(req, res, next) {
+  try {
+    const reports = await db.any('select * from report');
+    return res.status(200).json(reports);
+  } catch (err) {
+    return next(err);
+  }
 }
 
-function getSingleReport(req, res, next) {
-  var reportID = parseInt(req.params.id);
-  db
-    .one('select * from report where report_id = $1', reportID)
-    .then(function(data) {
-      res.status(200).json(data);
-    })
-    .catch(function(err) {
-      return next(err);
-    });
+async function getSingleReport(req, res, next) {
+  try {
+    const report = await getReportById(req.params.id);
+    return res.status(200).json(report);
+  } catch (err) {
+    return next(err);
+  }
 }
 
-function getAllreports_by_name(req, res, next) {
-  var name = req.query.name;
-  db
-    .any('select * from report where name = $1', name)
-    .then(function(data) {
-      res.status(200).json(data);
-    })
-    .catch(function(err) {
-      return next(err);
-    });
+async function getAllreports_by_name(req, res, next) {
+  try {
+    const reports = await db.any(
+      'select * from report where name = $1',
+      req.query.name
+    );
+    return res.status(200).json(reports);
+  } catch (err) {
+    return next(err);
+  }
 }
 
-function createReport(req, res, next) {
-  if (!req.body.hasOwnProperty('lims_entity')) {
-    // property may be null if it is not associated with a LIMS entity (Project, Library, Run, Pool)
-    req.body.lims_entity = null;
-  }
-  if (
-    !Object.values(req.body.permitted_parameters).every(p => {
-      const type = parseShesmuType(p.type);
-      return type != null && type.rest === '';
-    })
-  ) {
-    res.status(400).send('Invalid type signature on parameter.');
-  }
-  db
-    .one(
+function checkTypeValidity(parameters) {
```

Comment:
Can this be `checkTypeDefinitionValidity` to make it different from checking the values?